### PR TITLE
Download column order completed

### LINKED ISSUE #1062: Implement list of changes to export files
Implement all the changes outlined in "Indicators for UI and Export" and "Export file changes"

### PR DESCRIPTION
**Pull request by esfoobar-usds**
Fixes #1062

### CODE REVIEW

File: data/data-pipeline/data_pipeline/etl/score/constants.py [hunk 1]
```diff
@@ -194,88 +194,84 @@
...
+    field_names.EXPECTED_AGRICULTURE_LOSS_RATE_LOW_INCOME_FIELD,
     field_names.EXPECTED_AGRICULTURE_LOSS_RATE_FIELD
     + field_names.PERCENTILE_FIELD_SUFFIX,
-    field_names.EXPECTED_AGRICULTURE_LOSS_RATE_LOW_INCOME_FIELD,
-    field_names.EXPECTED_BUILDING_LOSS_RATE_FIELD,
+    field_names.EXPECTED_AGRICULTURE_LOSS_RATE_FIELD,
+    field_names.EXPECTED_BUILDING_LOSS_RATE_LOW_INCOME_FIELD,
     field_names.EXPECTED_BUILDING_LOSS_RATE_FIELD
     + field_names.PERCENTILE_FIELD_SUFFIX,
-    field_names.EXPECTED_BUILDING_LOSS_RATE_LOW_INCOME_FIELD,
```
Comment (lucasmbrown-usds) [inline data/data-pipeline/data_pipeline/etl/score/constants.py:207]: I think we're now missing `field_names.EXPECTED_BUILDING_LOSS_RATE_FIELD`?

Comment (lucasmbrown-usds) [reply data/data-pipeline/data_pipeline/etl/score/constants.py:207]: Oh, it looks like it's below on line 207. I think the building loss rate should be clustered with the other building loss rate fields here.

File: data/data-pipeline/data_pipeline/etl/score/constants.py [hunk 2]
```diff
@@ -194,88 +194,84 @@
...
-    field_names.EXPECTED_POPULATION_LOSS_RATE_FIELD,
+    field_names.EXPECTED_POPULATION_LOSS_RATE_LOW_INCOME_FIELD,
     field_names.EXPECTED_POPULATION_LOSS_RATE_FIELD
     + field_names.PERCENTILE_FIELD_SUFFIX,
```
Comment (lucasmbrown-usds) [inline data/data-pipeline/data_pipeline/etl/score/constants.py:206]: Seems like we're missing `field_names.EXPECTED_POPULATION_LOSS_RATE_FIELD`. 

I think every field needs to have: 

* BLAH_LOW_INCOME_FIELD (or for the workforce category, BLAH_LOW_HS_eDUCATION_FIELD) 
* BLAH + field_names.PERCENTILE_FIELD
* BLAH (the raw value)

File: data/data-pipeline/data_pipeline/etl/score/constants.py [hunk 3]
```diff
@@ -194,88 +194,84 @@
...
-    field_names.UNEMPLOYMENT_LOW_HS_EDUCATION_FIELD,
+    field_names.HIGH_SCHOOL_ED_FIELD,
     field_names.COMBINED_UNEMPLOYMENT_2010,
-    field_names.CENSUS_DECENNIAL_UNEMPLOYMENT_FIELD_2009,
     field_names.COMBINED_POVERTY_LESS_THAN_100_FPL_FIELD_2010,
     field_names.ISLAND_AREAS_UNEMPLOYMENT_LOW_HS_EDUCATION_FIELD,
     field_names.ISLAND_AREAS_POVERTY_LOW_HS_EDUCATION_FIELD,
     field_names.ISLAND_AREAS_LOW_MEDIAN_INCOME_LOW_HS_EDUCATION_FIELD,
-    field_names.ISLAND_AREAS_LOW_HS_EDUCATION_FIELD,
+    field_names.THRESHOLD_COUNT,
```
Comment (lucasmbrown-usds) [inline data/data-pipeline/data_pipeline/etl/score/constants.py:276]: I'm not exactly sure where this field is supposed to go, but I'm 90% sure it's supposed to be near the beginning of the column order.

File: data/data-pipeline/data_pipeline/etl/score/etl_score_post.py [hunk 4]
```diff
@@ -235,9 +235,11 @@ def _create_tile_data(
     def _create_downloadable_data(
         self, score_county_state_merged_df: pd.DataFrame
     ) -> pd.DataFrame:
-        return score_county_state_merged_df[
-            constants.DOWNLOADABLE_SCORE_COLUMNS
-        ]
+        df = score_county_state_merged_df[constants.DOWNLOADABLE_SCORE_COLUMNS]
+
+        # rename fields
```
Comment (lucasmbrown-usds) [inline data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:240]: remove?

**Review by lucasmbrown-usds — CHANGES_REQUESTED** [review]

File: data/data-pipeline/data_pipeline/etl/score/etl_score_post.py (hunk 4 above)
Comment (esfoobar-usds) [reply data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:240]: No, it's coming soon :) -- just for the GEOID10_TRACT field

File: data/data-pipeline/data_pipeline/etl/score/constants.py (hunk 1 above)
Comment (esfoobar-usds) [reply data/data-pipeline/data_pipeline/etl/score/constants.py:207]: Done

File: data/data-pipeline/data_pipeline/etl/score/constants.py (hunk 2 above)
Comment (esfoobar-usds) [reply data/data-pipeline/data_pipeline/etl/score/constants.py:206]: Done.

File: data/data-pipeline/data_pipeline/etl/score/constants.py (hunk 3 above)
Comment (esfoobar-usds) [reply data/data-pipeline/data_pipeline/etl/score/constants.py:276]: Done.

File: data/data-pipeline/data_pipeline/score/field_names.py [hunk 5]
```diff
@@ -106,16 +106,20 @@
...
 
 # Health
-DIABETES_FIELD = "Diagnosed diabetes among adults aged >=18 years"
-ASTHMA_FIELD = "Current asthma among adults aged >=18 years"
-HEART_DISEASE_FIELD = "Coronary heart disease among adults aged >=18 years"
-CANCER_FIELD = "Cancer (excluding skin cancer) among adults aged >=18 years"
+DIABETES_FIELD = (
+    "Diagnosed diabetes among adults aged greater than or equal to 18 years"
+)
+ASTHMA_FIELD = (
```
Comment (lucasmbrown-usds) [inline data/data-pipeline/data_pipeline/score/field_names.py:112]: Posting here as well as in our other conversation to discuss the bug on the field name: 

> this bug was probably introduced by this PR: https://github.com/usds/justice40-tool/pull/1077/files#diff-cc9c42ab3dd9470697a6a2f7b749bb4eef53337ab0c72b10b8669af1e8860396R112.
we rewrote the word “>=” into “greater than or equal to” for a bunch of fields, but some of these come directly from the source data without renaming.
i’d cast a strong preference for option #1 (rename it in the ETL script) out of the two you listed – i’m fairly nervous about introducing an additional renaming in a new place, i think that makes it hard for a developer to understand where different fields are coming from

File: data/data-pipeline/data_pipeline/etl/score/etl_score_post.py [hunk 6]
```diff
@@ -240,22 +242,21 @@ def _create_tile_data(
...
+        df = score_county_state_merged_df[
+            constants.DOWNLOADABLE_SCORE_COLUMNS
+        ].copy(deep=True)
+
+        df_of_float_columns = df.select_dtypes(include=["float64"])
+
+        for column in df_of_float_columns.columns:
+            for item in constants.PERCENT_PREFIXES_SUFFIXES:
+                if item in column:
+                    df[column] = df[column] * 100
```
Comment (esfoobar-usds) [inline data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:254]: @lucasmbrown-usds this is where we're trying to add code to cast as an integer if the value is not NaN, othewise leave as NaN

Comment (saran-ahluwalia) [reply data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:254]: @esfoobar-usds This is from @lucasmbrown-usds in Slack  `FYI we can also just multiply the percentiles by 100 when we create them in data/data-pipeline/data_pipeline/etl/score/etl_score.py:494`

Comment (lucasmbrown-usds) [reply data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:254]: Should just be: 

```
from data_pipeline.etl.score.etl_utils import floor_series

...

df[column] = (floor_series(
                series=df[column] * 100,
                number_of_decimals=0,
            )).astype('Int64')
```

Comment (saran-ahluwalia) [reply data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:254]: Yeah, we referenced that [nullable](https://pandas.pydata.org/pandas-docs/version/0.24/user_guide/integer_na.html) integer type as you showed Lucas. I think we were still have side effects propagate in some of the columns. I can retry that again and see if that does indeed rectify some of the problems we noted.

Comment (esfoobar-usds) [reply data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:254]: Got an error:

```
Exception has occurred: TypeError
data type 'Int64' not understood
  File "/opt/justice40-tool/data/data-pipeline/data_pipeline/etl/score/etl_score_post.py", line 253, in _create_downloadable_data
    df[column] = (
  File "/opt/justice40-tool/data/data-pipeline/data_pipeline/etl/score/etl_score_post.py", line 282, in transform
    self.output_downloadable_df = self._create_downloadable_data(
  File "/opt/justice40-tool/data/data-pipeline/data_pipeline/etl/runner.py", line 102, in score_post
    score_post.transform()
  File "/opt/justice40-tool/data/data-pipeline/data_pipeline/application.py", line 206, in generate_score_post
    score_post(data_source)
  File "/opt/justice40-tool/data/data-pipeline/data_pipeline/application.py", line 283, in <module>
    cli()
```

with this code:

```
                if item in column:
                    df[column] = (
                        floor_series(
                            series=df[column] * 100,
                            number_of_decimals=0,
                        )
                    ).astype("Int64")
```

File: data/data-pipeline/data_pipeline/etl/score/etl_score_post.py [hunk 7]
```diff
@@ -240,22 +242,21 @@ def _create_tile_data(
...
-        #     axis=0,
-        # )
-
-        # [x for x in df.columns if field_names.PERCENTILE_FIELD_SUFFIX in x]
-
+        df = score_county_state_merged_df[
+            constants.DOWNLOADABLE_SCORE_COLUMNS
+        ].copy(deep=True)
+
+        df_of_float_columns = df.select_dtypes(include=["float64"])
```
Comment (saran-ahluwalia) [inline data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:249]: I was thinking about this - we should just reference `df`'s columns as we verified that the input are float columns - although the nullable types causes indirection.

**Comment by esfoobar-usds** [comment]
I still see percentages with one zero decimal:

```
df[Expected agricultural loss rate (Natural Hazards Risk Index) (percentile)']
0        65.0
1        63.0
2        64.0
3        53.0
4        50.0
```

Not only that, non-percentiles have now one decimal as well?

```
df['Expected building loss rate (Natural Hazards Risk Index)']
0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
```

**Comment by esfoobar-usds** [comment]
I tried this:

```
            if any(x in column for x in constants.PERCENT_PREFIXES_SUFFIXES):
                # Convert percentages from fractions between 0 and 1 to an integer
                # from 0 to 100.
                df_100 = df[column] * 100
                df_int = pd.to_numeric(
                    arg=df_100,
                    errors="raise",
                    # Using `to_numeric` and downcast instead of `.astype(int)`
                    # because Python's `int` type does not support nulls.
                    downcast="integer",
                )
                print(df_100, df_int)
```

And `df_int` is identical to `df_`100`:

```
df_100:
0        65.280724
1        63.485934
2        64.312297
3        53.341470
4        50.406179
           ...    
74111    73.214214
74112          NaN
74113    75.901397
74114    74.592821
74115    74.462764
Name: Expected agricultural loss rate (Natural Hazards Risk Index) (percentile), Length: 74116, dtype: float64 0        65.280724

df_int
1        63.485934
2        64.312297
3        53.341470
4        50.406179
           ...    
74111    73.214214
74112          NaN
74113    75.901397
74114    74.592821
74115    74.462764
Name: Expected agricultural loss rate (Natural Hazards Risk Index) (percentile), Length: 74116, dtype: float64
```

**Comment by esfoobar-usds** [comment]
Ah! I WIN! 🎉 

```
            if any(x in column for x in constants.PERCENT_PREFIXES_SUFFIXES):
                # Convert percentages from fractions between 0 and 1 to an integer
                # from 0 to 100.
                df_100 = df[column] * 100
                df_int = np.floor(
                    pd.to_numeric(df_100, errors="coerce")
                ).astype("Int64")
                print(df_100, df_int)
```

```
df_100:
0        65.280724
1        63.485934
2        64.312297
3        53.341470
4        50.406179
           ...    
74111    73.214214
74112          NaN
74113    75.901397
74114    74.592821
74115    74.462764
Name: Expected agricultural loss rate (Natural Hazards Risk Index) (percentile), Length: 74116, dtype: float64 

df_int:
0          65
1          63
2          64
3          53
4          50
         ... 
74111      73
74112    <NA>
74113      75
74114      74
74115      74
Name: Expected agricultural loss rate (Natural Hazards Risk Index) (percentile), Length: 74116, dtype: Int64
```

The only weird thing is that `<NA>` value -- remains to see if this is expressed as null in the Excel sheet and none in the CSV.

File: data/data-pipeline/data_pipeline/etl/score/etl_score_post.py [hunk 8]
```diff
@@ -238,9 +242,31 @@ def _create_tile_data(
...
+            # TODO: create a schema for fields to make it more explicit and safe which
+            #  fields are percentages.
+            if any(x in column for x in constants.PERCENT_PREFIXES_SUFFIXES):
+                # Convert percentages from fractions between 0 and 1 to an integer
+                # from 0 to 100.
+                df_100 = df[column] * 100
+                df_int = np.floor(
+                    pd.to_numeric(df_100, errors="coerce")
+                ).astype("Int64")
+                df[column] = df_int
```
Comment (saran-ahluwalia) [inline data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:261]: @esfoobar-usds 

This is another way that simplifies the above code. See below for the output.

` df.loc[:, column] = (100 * df.loc[:, column]).round(0).astype("Int64")`

Screen Shot 2022-01-13 at 8 35 00 AM

File: data/data-pipeline/data_pipeline/etl/score/etl_score_post.py [hunk 9]
```diff
@@ -238,9 +242,31 @@ def _create_tile_data(
...
+            else:
+                # Round all other floats.
+                df[column] = floor_series(
```
Comment (saran-ahluwalia) [inline data/data-pipeline/data_pipeline/etl/score/etl_score_post.py:264]: 🥳

**Comment by esfoobar-usds** [comment]
> Overall looks good! TBH I did not check the order of the individual columns -- I've looked at these things so many times it's hard to follow, so I assume we'll find out if there's a problem doing QA on the actual file.

Totally -- I dropped a zip file on Slack.
